### ISSUE TITLE
Messages not received when DomainParticipant has custom listener

### ISSUE THREAD

**Issue by JakubSosnovec** [eProsima/Fast-DDS#1902]
Subscriber side in the example DDSHelloWorldExample is not receiving user messages when DomainParticipant object is created with custom listener. This is happening regardless of the transport layer used. Note that both participant and endpoint discovery and matching is still successful.

## Expected Behavior
DDSHelloWorldExample on subscriber side prints "Messages HelloWorld X RECEIVED" 

## Current Behavior
DDSHelloWorldExample on subscriber side doesnt print any user messages. 

## Steps to Reproduce
Here is the diff in DDSHelloWorldExample to reproduce the problem:

```
diff --git a/examples/C++/DDS/HelloWorldExample/HelloWorldSubscriber.cpp b/examples/C++/DDS/HelloWorldExample/HelloWorldSubscriber.cpp
index bc899fe29..a2b283b53 100644
--- a/examples/C++/DDS/HelloWorldExample/HelloWorldSubscriber.cpp
+++ b/examples/C++/DDS/HelloWorldExample/HelloWorldSubscriber.cpp
@@ -25,9 +25,16 @@
 #include <fastdds/dds/subscriber/DataReader.hpp>
 #include <fastdds/dds/subscriber/SampleInfo.hpp>
 #include <fastdds/dds/subscriber/qos/DataReaderQos.hpp>
+#include <fastdds/dds/domain/DomainParticipantListener.hpp>
 
 using namespace eprosima::fastdds::dds;
 
+class SubCustomDomainParticipantListener
+    : public eprosima::fastdds::dds::DomainParticipantListener {
+public:
+  virtual ~SubCustomDomainParticipantListener() {}
+};
+
 HelloWorldSubscriber::HelloWorldSubscriber()
     : participant_(nullptr)
     , subscriber_(nullptr)
@@ -41,7 +48,8 @@ bool HelloWorldSubscriber::init()
 {
     DomainParticipantQos pqos;
     pqos.name("Participant_sub");
-    participant_ = DomainParticipantFactory::get_instance()->create_participant(0, pqos);
+    SubCustomDomainParticipantListener *listener = new SubCustomDomainParticipantListener;
+    participant_ = DomainParticipantFactory::get_instance()->create_participant(0, pqos, listener);
 
     if (participant_ == nullptr)
     {

```

**Comment by MiguelCompany** [comment]
@JakubSosnovec 

**TL/DR:** If you are only interested in discovery events, pass `StatusMask::none()` as the fourth parameter of `create_participant()`

**Deep explanation:**
The `create_participant` method receives a fourth parameter with a status mask, indicating the status list for which the listener should be called. The default value for that parameter is `StatusMask::all()` meaning it should be called for changes on every kind of status.

There is a particular status (`data_on_readers`) which is triggered when any DataReader has a new sample ready to be returned. As stated on point 2.2.4.3.2 of the DDS standard, when a new sample is ready, the listeners for `data_on_readers` are triggered first. If the middleware cannot trigger an `on_data_on_readers` callback, it will then try to trigger `on_data_available`.

This means that on the code you posted, the default (empty) implementation of `DomainParticipantListener::on_data_on_readers` is being called.

**Comment by JakubSosnovec** [comment]
@MiguelCompany Thanks a lot for the prompt answer and the thorough explanation! Closing the ticket.

**Comment by Craig-Trimble** [comment]
Thanks, I missed that bit of the api :)
